agentliveness/agent.py

```diff
@@ -212,7 +212,47 @@ def _check_ironic(self):
 
         return 0
 
+    def _get_session_with_token_cache(self):
+        """Read the token from file and see if it still works
+
+        Returns `None` if the token didn't exist or expired.
+        """
+        try:
+            with open(self.CONF.token_cache_file, 'rb') as f:
+                import pickle
+                auth_ref = pickle.load(f)
```

Comment:
More an FIY, I'd say it, when going down the pickle road, this could be handled more elegantly with [Shelve](https://docs.python.org/3/library/shelve.html). One could open shelve in the beginning, store the token in a dict-like object, and only write it back on termination.

Reply:
I didn't know about `shelve`. I've switched to using it here. Thanks.